### PR TITLE
qt@5.9 (new formula)

### PR DESCRIPTION
**Pull request by apjanke**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting? *No. This is a beast of a build, so I'm letting the CI boxes handle it, so the discussion can proceed faster.*
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>`?

-----

Adds a qt@5.9 LTS formula, following discussion in https://github.com/Homebrew/homebrew-core/issues/35883#issuecomment-455101919.

Passes `brew audit --strict` locally, but I haven't done a build yet due to the time involved; I'm going to lean on the CI servers for that. Sorry if that's being presumptuous.

### CODE REVIEW

File: Formula/qt@5.9.rb [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+  keg_only :versioned_formula
+
+  option "with-docs", "Build documentation"
+  option "with-examples", "Build examples"
+
+  depends_on "pkg-config" => :build
+  depends_on :xcode => :build
+  depends_on "mysql" => :optional
+  depends_on "postgresql" => :optional
```
Comment (MikeMcQuaid) [inline Formula/qt@5.9.rb]: Remove all options.

File: Formula/qt@5.9.rb [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+  # Restore `.pc` files for framework-based build of Qt 5 on OS X. This
+  # partially reverts <https://codereview.qt-project.org/#/c/140954/> merged
+  # between the 5.5.1 and 5.6.0 releases. (Remove this as soon as feasible!)
+  #
+  # Core formulae known to fail without this patch (as of 2016-10-15):
+  #   * gnuplot  (with `--with-qt` option)
+  #   * mkvtoolnix (with `--with-qt` option, silent build failure)
+  #   * poppler    (with `--with-qt` option)
+  patch do
```
Comment (MikeMcQuaid) [inline Formula/qt@5.9.rb]: What's the scope on either of these patches being merged upstream?

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

File: Formula/qt@5.9.rb (hunk 2 above)
Comment (apjanke) [reply Formula/qt@5.9.rb]: The first patch... is a long answer. I've spent days in the weeds on this with Octave recently and I still don't know. (That is, I'm still not clear on what the exact issue is, and whether it's landed already.)

The second patch got landed in Qt 5.10; I don't think it'll be backported to the 5.9 series.

File: Formula/qt@5.9.rb (hunk 1 above)
Comment (apjanke) [reply Formula/qt@5.9.rb]: Options removed.

File: Formula/qt@5.9.rb [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+class QtAT59 < Formula
+  desc "Cross-platform application and UI framework, Long-Term Support version"
+  homepage "https://www.qt.io/"
+  url "https://download.qt.io/official_releases/qt/5.9/5.9.3/single/qt-everywhere-opensource-src-5.9.3.tar.xz"
+  mirror "https://www.mirrorservice.org/sites/download.qt-project.org/official_releases/qt/5.9/5.9.3/single/qt-everywhere-opensource-src-5.9.3.tar.xz"
+  sha256 "57acd8f03f830c2d7dc29fbe28aaa96781b2b9bdddce94196e6761a0f88c6046"
+
+  keg_only :versioned_formula
+
+  option "with-docs", "Build documentation"
```
Comment (fxcoudert) [inline Formula/qt@5.9.rb]: Still options :)

Comment (apjanke) [reply Formula/qt@5.9.rb]: *smacks forehead, removes remaining options*

File: Formula/qt@5.9.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Formula/qt@5.9.rb]: @apjanke I think we should aim to include this without these patches unless it's useless without. At the very least would be good to request/attempt a backport for these if it's an LTS.

Comment (Bo98) [reply Formula/qt@5.9.rb]: First patch was landed in 5.12. I'm not aware of a bug report for it however.

Comment (MikeMcQuaid) [reply Formula/qt@5.9.rb]: Backports for this probably require using their Gerrit (https://codereview.qt-project.org). Would like to see them requested before this is merged.

**Comment by tooomm** [comment]
Any news or progress lately?

**Comment by apjanke** [comment]
> Any news or progress lately?

Sorry; busy with the paying job. :) I need to ping/submit upstream to see what they say about backporting this.

**Comment by MikeMcQuaid** [comment]
> I need to ping/submit upstream to see what they say about backporting this.

@apjanke if you detail what needs to be done someone else might pick it up and beat you to it?

**Comment by fxcoudert** [comment]
My suggestion:

- Patch #1 (`Restore .pc files`) can go away. It was needed for building some formulas from source, but in most use cases it is not (as I understand it) necessary.
- For patch #2, I requested a backport on 18 Jan 2019: https://bugreports.qt.io/browse/QTBUG-56814, so I would say it is acceptable.

**Comment by MikeMcQuaid** [comment]
> _Edit: I've now requested the backport on Gerrit as well: https://codereview.qt-project.org/#/c/180825/_
> 
> @MikeMcQuaid @apjanke if you agree with above, I can finish this

Agreed :+1:

**Comment by apjanke** [comment]
Also agreed. Go for it: I am short on time right now.

File: Formula/qt@5.9.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  mirror "https://www.mirrorservice.org/sites/download.qt-project.org/official_releases/qt/5.9/5.9.3/single/qt-everywhere-opensource-src-5.9.3.tar.xz"
+  sha256 "57acd8f03f830c2d7dc29fbe28aaa96781b2b9bdddce94196e6761a0f88c6046"
+
+  keg_only :versioned_formula
+
+  depends_on "pkg-config" => :build
+  depends_on :xcode => :build
+
+  # Fix for upstream issue "macdeployqt does not work with Homebrew"
+  # See https://bugreports.qt.io/browse/QTBUG-56814
```
Comment (MikeMcQuaid) [inline Formula/qt@5.9.rb]: Could you link the Gerrit here?

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by fxcoudert** [comment]
Now it's not building anymore:

```
mac/qfontengine_coretext.mm:833:20: error: qualified reference to 'QFixed' is a constructor name rather than a type in this context
    return QFixed::QFixed(int(CTFontGetUnitsPerEm(ctfont)));
                   ^
```

**Comment by fxcoudert** [comment]
Bumping to 5.9.7. There are no known issues there that should affect us: https://wiki.qt.io/Qt_5.9.7_Known_Issues

File: Formula/qt@5.9.rb [hunk 5]
```diff
@@ -7,17 +7,17 @@
 class QtAT59 < Formula
   desc "Cross-platform application and UI framework, Long-Term Support version"
   homepage "https://www.qt.io/"
-  url "https://download.qt.io/official_releases/qt/5.9/5.9.3/single/qt-everywhere-opensource-src-5.9.3.tar.xz"
-  mirror "https://www.mirrorservice.org/sites/download.qt-project.org/official_releases/qt/5.9/5.9.3/single/qt-everywhere-opensource-src-5.9.3.tar.xz"
-  sha256 "57acd8f03f830c2d7dc29fbe28aaa96781b2b9bdddce94196e6761a0f88c6046"
+  url "https://download.qt.io/official_releases/qt/5.9/5.9.7/single/qt-everywhere-opensource-src-5.9.7.tar.xz"
+  mirror "https://www.mirrorservice.org/sites/download.qt-project.org/official_releases/qt/5.9/5.9.7/single/qt-everywhere-opensource-src-5.9.7.tar.xz"
+  sha256 "70e617aeb1f9bbf84a12b8cf09b01ece"
```
Comment (Bo98) [inline Formula/qt@5.9.rb]: That's the MD5 :)

Think we're looking for `1c3852aa48b5a1310108382fb8f6185560cefc3802e81ecc099f4e62ee38516c`

Comment (fxcoudert) [reply Formula/qt@5.9.rb]: This shows just how tired I am 😭

**Comment by fxcoudert** [comment]
@BrewTestBot test this please

**Comment by Bo98** [comment]
Looks like an incompatibility with the bundled version of Boost and C++17.

I suppose there's a few approaches to fix it (using our version of Boost, supplying a compiler option to re-enable auto_ptr, ...) but maybe switching to C++14 will be easiest.

**Comment by fxcoudert** [comment]
Well, sadly if the bundled boost from Qt 5.9 is too old to build as-is, I guess that means it's not suitable for us. I'm closing for now, any maintainer feel free to reopen with a solution to that issue, if you wish :)

**Pull request closed without merging**